pkg/controller/sparkapplication/controller.go

```diff
@@ -289,397 +298,464 @@ func (c *Controller) processNextItem() bool {
 	}
+	var command = filepath.Join(sparkHome, "/bin/spark-submit")
 
-	c.recordExecutorEvent(app, update.state, update.podName)
+	cmd := execCommand(command, submission.args...)
+	glog.Infof("spark-submit arguments: %v", cmd.Args)
 
-	return c.updateSparkApplicationStatusWithRetries(app, func(status *v1alpha1.SparkApplicationStatus) {
-		if status.ExecutorState == nil {
-			status.ExecutorState = make(map[string]v1alpha1.ExecutorState)
+	if _, err := cmd.Output(); err != nil {
+		var errorMsg string
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			errorMsg = string(exitErr.Stderr)
 		}
-		existingState, ok := status.ExecutorState[update.podName]
-		if ok && isExecutorTerminated(existingState) {
-			return
+		// Already Exists. Do nothing.
+		if strings.Contains(errorMsg, podAlreadyExistsErrorCode) {
+			glog.Warningf("Trying to resubmit an already submitted SparkApplication %s in namespace %s. Error: %v", submission.name, submission.namespace, errorMsg)
+			return nil
 		}
-		status.ExecutorState[update.podName] = update.state
-	})
-}
-
-func (c *Controller) updateSparkApplicationStatusWithRetries(
-	original *v1alpha1.SparkApplication,
-	updateFunc func(*v1alpha1.SparkApplicationStatus)) *v1alpha1.SparkApplication {
-	var lastUpdateErr error
-	for i := 0; i < maximumUpdateRetries; i++ {
-		toUpdate := original.DeepCopy()
-		glog.V(2).Infof("Trying to update SparkApplication %s", original.Name)
-		// Apply update
-		updateFunc(&toUpdate.Status)
+		glog.Errorf("failed to run spark-submit for SparkApplication %s in namespace: %s. Error: %v", submission.name,
+			submission.namespace, errorMsg)
+		app.Status = v1alpha1.SparkApplicationStatus{
+			AppState: v1alpha1.ApplicationState{
+				State:        v1alpha1.FailedSubmissionState,
+				ErrorMessage: errorMsg,
+			},
+			SubmissionAttempts: app.Status.SubmissionAttempts + 1,
+			SubmissionTime:     metav1.Now(),
+		}
+		c.recordSparkApplicationEvent(app)
+	} else {
+		glog.Infof("spark-submit completed for SparkApplication %s in namespace %s", submission.name, submission.namespace)
+
+		// Update AppStatus to submitted.
+		app.Status = v1alpha1.SparkApplicationStatus{
+			AppState: v1alpha1.ApplicationState{
+				State: v1alpha1.SubmittedState,
+			},
+			SubmissionAttempts: app.Status.SubmissionAttempts + 1,
+			ExecutionAttempts:  app.Status.ExecutionAttempts + 1,
+			SubmissionTime:     metav1.Now(),
+		}
+		c.recordSparkApplicationEvent(app)
 
-		// Let's keep the old App status if this is not a valid transition.
-		if original.Status.AppState.State != toUpdate.Status.AppState.State &&
-			!isValidDriverStateTransition(original.Status.AppState.State, toUpdate.Status.AppState.State) {
-			glog.Warningf("Invalid Driver State Transition. From:[%v] To:[%v]", original.Status.AppState.State, toUpdate.Status.AppState.State)
-			toUpdate.Status.AppState = original.Status.AppState
+		// Add driver as a finalizer to prevent SparkApplication deletion till driver is deleted.
```

Comment:
Can we put this into a utility method corresponding to `remoteFinalizer` and put it into `sparkapp_util.go`?